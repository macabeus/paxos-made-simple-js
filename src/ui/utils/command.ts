export type ParsedCommand =
  | { action: "invalid"; reason: string }
  | { action: "quit" }
  | { action: "help" }
  | { action: "propose"; worker: string; value: string }
  | { action: "add-worker"; id: string }
  | { action: "delay"; miliseconds: number }
  | { action: "off"; id: string }
  | { action: "on"; id: string };

const commandParser = (raw: string): ParsedCommand => {
  const parts = raw.split(" ");

  if (parts[0] === "q" || parts[0] === "quit") {
    return { action: "quit" };
  }

  if (parts[0] === "h" || parts[0] === "help") {
    return { action: "help" };
  }

  if (parts[0] === "p" || parts[0] === "propose") {
    if (parts.length !== 3) {
      return {
        action: "invalid",
        reason:
          '"proposal" command needs exactly two parameters: `proposal workerId value`',
      };
    }

    return { action: "propose", worker: parts[1], value: parts[2] };
  }

  if (parts[0] === "a" || parts[0] === "add-worker") {
    if (parts.length !== 2) {
      return {
        action: "invalid",
        reason:
          '"add-worker" command needs exactly one parameter: `add-worker workerId`',
      };
    }

    return { action: "add-worker", id: parts[1] };
  }

  if (parts[0] === "d" || parts[0] === "delay") {
    if (parts.length !== 2) {
      return {
        action: "invalid",
        reason:
          '"delay" command needs exactly one parameter: `delay miliseconds`',
      };
    }

    return { action: "delay", miliseconds: Number(parts[1]) };
  }

  if (parts[0] === "off") {
    if (parts.length !== 2) {
      return {
        action: "invalid",
        reason: '"off" command needs exactly one parameter: `off workerId`',
      };
    }

    return { action: "off", id: parts[1] };
  }

  if (parts[0] === "on") {
    if (parts.length !== 2) {
      return {
        action: "invalid",
        reason: '"on" command needs exactly one parameter: `on workerId`',
      };
    }

    return { action: "on", id: parts[1] };
  }

  return { action: "invalid", reason: `Input "${raw}" is not a valid command` };
};

export const commandParserMultiples = (raw: string): ParsedCommand[] => {
  const rawParts = raw.split(/;\s*/);

  const parsedCommands = rawParts.map((raw) => commandParser(raw));

  const includeInvalidCommand = parsedCommands.find(
    (command) => command.action === "invalid"
  );
  if (includeInvalidCommand) {
    return [includeInvalidCommand];
  }

  return parsedCommands;
};
