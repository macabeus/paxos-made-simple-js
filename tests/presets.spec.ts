import { describe, it } from "vitest";
import { loadPreset } from "./testUtils";
import { delay } from "../src/ui/utils/helpers";

describe.concurrent("# Run Presets", () => {
  it("Simple", async ({ expect }) => {
    const { runDefaultCommand, getStatus } = await loadPreset("Simple");

    await runDefaultCommand();
    await delay(25_000);

    const status = getStatus();
    expect(status).matchSnapshot();
  });

  it("Trying proposing with an obsoleted id", async ({ expect }) => {
    const { runDefaultCommand, getStatus } = await loadPreset(
      "Trying proposing with an obsoleted id"
    );

    await runDefaultCommand();
    await delay(15_000);

    const status = getStatus();
    expect(status).matchSnapshot();
  });

  it("Two proposes at the same time", async ({ expect }) => {
    const { runDefaultCommand, getStatus } = await loadPreset(
      "Two proposes at the same time"
    );

    await runDefaultCommand();
    await delay(25_000);

    const status = getStatus();
    expect(status).matchSnapshot();
  });

  it("Three proposes at the same time", async ({ expect }) => {
    const { runDefaultCommand, getStatus } = await loadPreset(
      "Three proposes at the same time"
    );

    await runDefaultCommand();
    await delay(30_000);

    const status = getStatus();
    expect(status).matchSnapshot();
  });

  it("Multiple proposes running sequentially", async ({ expect }) => {
    const { runDefaultCommand, getStatus } = await loadPreset(
      "Multiple proposes running sequentially"
    );

    await runDefaultCommand();
    await delay(25_000);

    const status = getStatus();
    expect(status).matchSnapshot();
  });

  it("Two proposes proposing the same value at the same time", async ({
    expect,
  }) => {
    const { runDefaultCommand, getStatus } = await loadPreset(
      "Two proposes proposing the same value at the same time"
    );

    await runDefaultCommand();
    await delay(25_000);

    const status = getStatus();
    expect(status).matchSnapshot();
  });

  it("Always without a majority", async ({ expect }) => {
    const { runDefaultCommand, getStatus } = await loadPreset(
      "Always without a majority"
    );

    await runDefaultCommand();
    await delay(25_000);

    const status = getStatus();
    expect(status).matchSnapshot();
  });

  it("No online majority initially", async ({ expect }) => {
    const { runDefaultCommand, getStatus } = await loadPreset(
      "No online majority initially"
    );

    await runDefaultCommand();
    await delay(25_000);

    const status = getStatus();
    expect(status).matchSnapshot();
  });
});
