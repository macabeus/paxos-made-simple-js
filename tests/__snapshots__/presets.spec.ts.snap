// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`# Run Presets > Always without a majority 1`] = `
{
  "worker1": {
    "acceptsReceived": 0,
    "minimumQuorum": 2,
    "proposingId": "Always without a majority#worker1:1",
    "proposingValue": "foo",
    "status": "proposing",
  },
  "worker2": {
    "highestPromiseId": "Always without a majority#worker1:1",
    "minimumQuorum": 2,
    "status": "promiseSent",
  },
  "worker3": {
    "minimumQuorum": 2,
    "status": "off",
  },
  "worker4": {
    "minimumQuorum": 2,
    "status": "off",
  },
}
`;

exports[`# Run Presets > Multiple proposes running sequentially 1`] = `
{
  "worker1": {
    "acceptedValue": "foo",
    "acceptsReceived": 2,
    "highestAcceptedId": "Multiple proposes running sequentially#worker1:1",
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker2": {
    "acceptedValue": "foo",
    "acceptsReceived": 0,
    "highestAcceptedId": "Multiple proposes running sequentially#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker3": {
    "acceptedValue": "foo",
    "acceptsReceived": 0,
    "highestAcceptedId": "Multiple proposes running sequentially#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker4": {
    "acceptedValue": "foo",
    "highestAcceptedId": "Multiple proposes running sequentially#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker5": {
    "acceptedValue": "foo",
    "highestAcceptedId": "Multiple proposes running sequentially#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
}
`;

exports[`# Run Presets > No online majority initially 1`] = `
{
  "worker1": {
    "acceptedValue": "foo",
    "acceptsReceived": 2,
    "highestAcceptedId": "No online majority initially#worker1:1",
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker2": {
    "acceptedValue": "foo",
    "highestAcceptedId": "No online majority initially#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker3": {
    "acceptedValue": "foo",
    "acceptsReceived": 0,
    "highestAcceptedId": "No online majority initially#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker4": {
    "minimumQuorum": 2,
    "status": "off",
  },
}
`;

exports[`# Run Presets > Simple 1`] = `
{
  "worker1": {
    "acceptedValue": "foo",
    "acceptsReceived": 1,
    "highestAcceptedId": "Simple#worker1:1",
    "minimumQuorum": 1,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker2": {
    "acceptedValue": "foo",
    "highestAcceptedId": "Simple#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 1,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker3": {
    "acceptedValue": "foo",
    "highestAcceptedId": "Simple#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 1,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
}
`;

exports[`# Run Presets > Three proposes at the same time 1`] = `
{
  "worker1": {
    "acceptedValue": "foo",
    "acceptsReceived": 2,
    "highestAcceptedId": "Three proposes at the same time#worker1:1",
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker2": {
    "acceptedValue": "foo",
    "acceptsReceived": 0,
    "highestAcceptedId": "Three proposes at the same time#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker3": {
    "acceptedValue": "foo",
    "acceptsReceived": 0,
    "highestAcceptedId": "Three proposes at the same time#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker4": {
    "acceptedValue": "foo",
    "highestAcceptedId": "Three proposes at the same time#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker5": {
    "acceptedValue": "foo",
    "highestAcceptedId": "Three proposes at the same time#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
}
`;

exports[`# Run Presets > Trying proposing with an obsoleted id 1`] = `
{
  "worker1": {
    "acceptedValue": "foo",
    "acceptsReceived": 1,
    "highestAcceptedId": "Trying proposing with an obsoleted id#worker1:1",
    "minimumQuorum": 1,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker2": {
    "acceptedValue": "foo",
    "highestAcceptedId": "Trying proposing with an obsoleted id#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 1,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker3": {
    "acceptedValue": "foo",
    "acceptsReceived": 0,
    "highestAcceptedId": "Trying proposing with an obsoleted id#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 1,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
}
`;

exports[`# Run Presets > Two proposes at the same time 1`] = `
{
  "worker1": {
    "acceptedValue": "foo",
    "acceptsReceived": 2,
    "highestAcceptedId": "Two proposes at the same time#worker1:1",
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker2": {
    "acceptedValue": "foo",
    "acceptsReceived": 0,
    "highestAcceptedId": "Two proposes at the same time#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker3": {
    "acceptedValue": "foo",
    "highestAcceptedId": "Two proposes at the same time#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker4": {
    "acceptedValue": "foo",
    "highestAcceptedId": "Two proposes at the same time#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
}
`;

exports[`# Run Presets > Two proposes proposing the same value at the same time 1`] = `
{
  "worker1": {
    "acceptedValue": "foo",
    "acceptsReceived": 2,
    "highestAcceptedId": "Two proposes proposing the same value at the same time#worker1:1",
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker2": {
    "acceptedValue": "foo",
    "acceptsReceived": 0,
    "highestAcceptedId": "Two proposes proposing the same value at the same time#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker3": {
    "acceptedValue": "foo",
    "highestAcceptedId": "Two proposes proposing the same value at the same time#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker4": {
    "acceptedValue": "foo",
    "highestAcceptedId": "Two proposes proposing the same value at the same time#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
  "worker5": {
    "acceptedValue": "foo",
    "highestAcceptedId": "Two proposes proposing the same value at the same time#worker1:1",
    "highestPromiseId": null,
    "minimumQuorum": 2,
    "proposingId": null,
    "proposingValue": null,
    "status": "idle",
  },
}
`;
